File: backend/app/models/stock.py
"""
주식 데이터 모델
"""
from __future__ import annotations
from pydantic import BaseModel
from typing import Optional, Dict, Any
from datetime import datetime


class PriceInfo(BaseModel):
    """가격 정보"""
    current: Optional[float] = None
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None  # 종가 추가
    volume: Optional[int] = None


class FinancialsInfo(BaseModel):
    """재무 지표"""
    # 밸류에이션
    trailing_pe: Optional[float] = None
    forward_pe: Optional[float] = None
    pbr: Optional[float] = None
    roe: Optional[float] = None
    opm: Optional[float] = None
    peg: Optional[float] = None

    # 재무 건전성
    debt_to_equity: Optional[float] = None
    current_ratio: Optional[float] = None
    quick_ratio: Optional[float] = None

    # 배당
    dividend_yield: Optional[float] = None
    payout_ratio: Optional[float] = None

    # 성장성
    revenue_growth: Optional[float] = None
    earnings_growth: Optional[float] = None


class CompanyInfo(BaseModel):
    """회사 정보"""
    name: Optional[str] = None
    sector: Optional[str] = None
    industry: Optional[str] = None
    summary_original: Optional[str] = None
    summary_translated: Optional[str] = None


class SMAInfo(BaseModel):
    """단순이동평균 (SMA)"""
    sma20: Optional[float] = None
    sma50: Optional[float] = None
    sma200: Optional[float] = None


class EMAInfo(BaseModel):
    """지수이동평균 (EMA)"""
    ema12: Optional[float] = None
    ema26: Optional[float] = None


class RSIInfo(BaseModel):
    """상대강도지수 (RSI)"""
    rsi14: Optional[float] = None


class MACDInfo(BaseModel):
    """MACD 지표"""
    macd: Optional[float] = None
    signal: Optional[float] = None
    histogram: Optional[float] = None


class BollingerBandsInfo(BaseModel):
    """볼린저밴드"""
    upper: Optional[float] = None
    middle: Optional[float] = None
    lower: Optional[float] = None


class TechnicalIndicators(BaseModel):
    """기술적 지표 통합"""
    sma: Optional[SMAInfo] = None
    ema: Optional[EMAInfo] = None
    rsi: Optional[RSIInfo] = None
    macd: Optional[MACDInfo] = None
    bollinger_bands: Optional[BollingerBandsInfo] = None


class NewsItem(BaseModel):
    """뉴스 기사"""
    title: str
    link: str
    published_at: Optional[datetime] = None
    source: Optional[str] = None


class AIAnalysis(BaseModel):
    """Gemini AI 종합 분석 결과"""
    report: str


class StockData(BaseModel):
    """주식 데이터"""
    ticker: str
    timestamp: datetime
    market_cap: Optional[float] = None
    price: PriceInfo
    financials: FinancialsInfo
    company: CompanyInfo
    technical_indicators: Optional[TechnicalIndicators] = None
    chart_data: Optional[list[ChartDataPoint]] = None  # 차트 데이터 추가
    news: Optional[list[NewsItem]] = None
    ai_analysis: Optional[AIAnalysis] = None


class StockResponse(BaseModel):
    """API 응답 모델"""
    success: bool
    data: Optional[StockData] = None
    error: Optional[str] = None

class HistoricalStockData(BaseModel):
    """과거 주식 데이터"""
    ticker: str
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: Optional[int] = None

class HistoricalStockResponse(BaseModel):
    """과거 주식 데이터 API 응답 모델"""
    success: bool
    data: Optional[HistoricalStockData] = None
    error: Optional[str] = None

class NewsResponse(BaseModel):
    """뉴스 API 응답 모델"""
    success: bool
    data: Optional[list[NewsItem]] = None
    error: Optional[str] = None


class AnalysisResponse(BaseModel):
    """AI 분석 API 응답 모델"""
    success: bool
    data: Optional[AIAnalysis] = None
    error: Optional[str] = None


class ChartDataPoint(BaseModel):
    """차트의 단일 데이터 포인트"""
    date: str
    close: Optional[float] = None
    volume: Optional[int] = None
    sma20: Optional[float] = None
    sma50: Optional[float] = None
    sma200: Optional[float] = None
    rsi: Optional[float] = None
    macd: Optional[float] = None
    macd_signal: Optional[float] = None
    macd_hist: Optional[float] = None
    bb_upper: Optional[float] = None
    bb_middle: Optional[float] = None
    bb_lower: Optional[float] = None


class ChartResponse(BaseModel):
    """차트 데이터 API 응답 모델"""
    success: bool
    data: Optional[list[ChartDataPoint]] = None
    error: Optional[str] = None

File: frontend/src/types/stock.ts
export interface StockPrice {
  current: number;
  open: number;
  high: number;
  low: number;
  close: number | null;  // 종가 추가
  volume: number;
}

export interface StockFinancials {
  trailing_pe: number | null;
  forward_pe: number | null;
  pbr: number | null;
  roe: number | null;
  opm: number | null;
  peg: number | null;
  debt_to_equity: number | null;
  current_ratio: number | null;
  quick_ratio: number | null;
  dividend_yield: number | null;
  payout_ratio: number | null;
  revenue_growth: number | null;
  earnings_growth: number | null;
}

export interface StockCompany {
  name: string;
  sector: string | null;
  industry: string | null;
  summary_original: string | null;
  summary_translated: string | null;
}

export interface SMAInfo {
  sma20: number | null;
  sma50: number | null;
  sma200: number | null;
}

export interface EMAInfo {
  ema12: number | null;
  ema26: number | null;
}

export interface RSIInfo {
  rsi14: number | null;
}

export interface MACDInfo {
  macd: number | null;
  signal: number | null;
  histogram: number | null;
}

export interface BollingerBandsInfo {
  upper: number | null;
  middle: number | null;
  lower: number | null;
}

export interface TechnicalIndicators {
  sma: SMAInfo | null;
  ema: EMAInfo | null;
  rsi: RSIInfo | null;
  macd: MACDInfo | null;
  bollinger_bands: BollingerBandsInfo | null;
}

export interface NewsItem {
  title: string;
  link: string;
  published_at: string | null;
  source: string | null;
}

export interface StockData {
  ticker: string;
  timestamp: string;
  market_cap: number | null;
  price: StockPrice;
  financials: StockFinancials;
  company: StockCompany;
  technical_indicators?: TechnicalIndicators | null;
  chart_data?: ChartDataPoint[] | null;
  news?: NewsItem[] | null;
}

export interface AIAnalysis {
  report: string;
}

export interface ChartDataPoint {
  date: string;
  close: number | null;
  volume: number | null;
  sma20: number | null;
  sma50: number | null;
  sma200: number | null;
  rsi: number | null;
  macd: number | null;
  macd_signal: number | null;
  macd_hist: number | null;
  bb_upper: number | null;
  bb_middle: number | null;
  bb_lower: number | null;
}

// ============ AI 분석 요약 저장 관련 타입 ============

export type InvestmentStrategy = 'buy' | 'hold' | 'sell';

export interface AnalysisSummary {
  summary: string;  // 3줄 요약 (줄바꿈으로 구분)
  strategy: InvestmentStrategy;
}

export interface SummaryRequest {
  full_report: string;
}

export interface SummaryResponse {
  success: boolean;
  data: AnalysisSummary | null;
  error: string | null;
}

export interface SavedAnalysis {
  id: number;
  ticker: string;
  summary: string;
  strategy: string;
  current_price: number | null;
  user_avg_price: number | null;
  profit_loss_ratio: number | null;
  full_report: string | null;
  created_at: string;
}

export interface SaveAnalysisRequest {
  summary: string;
  strategy: InvestmentStrategy;
  current_price?: number;
  user_avg_price?: number;
  profit_loss_ratio?: number;
  full_report?: string;
}

export interface SaveAnalysisResponse {
  success: boolean;
  data: SavedAnalysis | null;
  error: string | null;
}

export interface AnalysisListResponse {
  success: boolean;
  data: SavedAnalysis[] | null;
  error: string | null;
}
